File: src/services/timelineService.ts
import { TreatmentStage } from "../types";

export interface GenerateTimelineParams {
  startDate: string; // YYYY-MM-DD or readable string
  durationMonths: number;
  numberOfSittings: number;
  customSittingDates?: string[];
}

/**
 * Automatically calculates the treatment timeline according to doctor prescriptions.
 * Supports automatic spacing or custom sitting dates.
 */
export function generateTreatmentTimeline({
  startDate,
  durationMonths,
  numberOfSittings,
  customSittingDates,
}: GenerateTimelineParams): TreatmentStage[] {
  const parsedStartDate = new Date(startDate);
  const validStartDate = isNaN(parsedStartDate.getTime()) ? new Date() : parsedStartDate;

  const count = Math.max(1, numberOfSittings);
  const stages: TreatmentStage[] = [];

  // Stage 0: Initial
  stages.push({
    id: "stage-0",
    stageName: "Initial",
    month: 0,
    date: formatDate(validStartDate),
    sittingNumber: 0,
    progress: 0,
    status: "completed",
    clinicalObservations: "Baseline diagnostic records, facial & smile photography, 3D intraoral scan.",
  });

  // Intermediate and Final stages
  const monthStep = durationMonths / count;

  for (let i = 1; i <= count; i++) {
    const isFinal = i === count;
    const rawMonth = Math.round(i * monthStep);
    const month = Math.min(durationMonths, rawMonth);
    const progress = Math.min(100, Math.round((i / count) * 100));

    let stageDateStr: string;
    if (customSittingDates && customSittingDates[i - 1]) {
      stageDateStr = formatDate(new Date(customSittingDates[i - 1]));
    } else {
      const stageDate = new Date(validStartDate);
      stageDate.setMonth(stageDate.getMonth() + Math.round((i * durationMonths) / count));
      stageDateStr = formatDate(stageDate);
    }

    const stageName = isFinal ? `Final — Month ${durationMonths}` : `Month ${month}`;

    stages.push({
      id: `stage-${i}`,
      stageName,
      month,
      date: stageDateStr,
      sittingNumber: i,
      progress,
      status: "upcoming",
      clinicalObservations: isFinal
        ? "Target aesthetic and occlusal alignment achieved. Retention phase initiation."
        : `Progress check for Sitting ${i} of ${count}. Alignment tracking at ${progress}%.`,
    });
  }

  // Set the first non-initial stage as 'current' if pending
  if (stages.length > 1) {
    stages[1].status = "current";
  }

  return stages;
}

export function formatDate(date: Date): string {
  if (isNaN(date.getTime())) return "";
  return date.toLocaleDateString("en-US", {
    day: "numeric",
    month: "short",
    year: "numeric",
  });
}

export function addMonths(date: Date, months: number): Date {
  const d = new Date(date);
  d.setMonth(d.getMonth() + months);
  return d;
}

File: backend/app/services/timeline_service.py
from typing import List, Optional, Dict, Any, Union
from datetime import date
from app.utils.dates import parse_date, add_months_to_date, format_iso_date


def generate_treatment_timeline(
    start_date: Union[str, date],
    duration_months: int,
    number_of_sittings: int,
    custom_dates: Optional[List[Union[str, date]]] = None
) -> List[Dict[str, Any]]:
    """
    Automatically calculates the treatment timeline according to doctor prescriptions.
    Supports automatic spacing or custom sitting dates prescribed by the dentist.
    
    Generates:
    - Stage 0: Initial (Month 0, 0% progress)
    - Stages 1..N: Intermediate and Final stages (up to 100% progress)
    """
    parsed_start_date = parse_date(start_date)
    duration_months = max(1, duration_months)
    count = max(1, number_of_sittings)

    stages: List[Dict[str, Any]] = []

    # 1. Initial Stage (Stage 0)
    stages.append({
        "stage_number": 0,
        "stage_name": "Initial",
        "month": 0,
        "scheduled_date": format_iso_date(parsed_start_date),
        "progress_percentage": 0,
        "status": "completed",
        "dentist_notes": "Baseline diagnostic smile records and initial assessment.",
    })

    # 2. Subsequent Stages (Stages 1..count)
    month_step = duration_months / count

    for i in range(1, count + 1):
        is_final = (i == count)
        raw_month = round(i * month_step)
        month = min(duration_months, raw_month)
        
        # Calculate progress percentage (0-100)
        progress = min(100, round((i / count) * 100))

        # Determine scheduled date
        if custom_dates and len(custom_dates) >= i and custom_dates[i - 1]:
            scheduled = parse_date(custom_dates[i - 1])
        else:
            # Add calculated months from start date
            scheduled = add_months_to_date(parsed_start_date, round(i * duration_months / count))

        if is_final:
            stage_name = f"Final — Month {duration_months}"
        else:
            stage_name = f"Month {month}"

        stages.append({
            "stage_number": i,
            "stage_name": stage_name,
            "month": month,
            "scheduled_date": format_iso_date(scheduled),
            "progress_percentage": progress,
            "status": "upcoming",
            "dentist_notes": f"Treatment sitting {i} of {count}. Target progress {progress}%.",
        })

    return stages
